pyrit/orchestrator/single_turn/question_answer_benchmark_orchestrator.py

```
@@ -0,0 +1,163 @@
+            prompt_converters (list[PromptConverter], Optional): The prompt converters to be used.
+            evaluation_prompt (str, Optional): The evaluation prompt to be used. Defaults to None.
+            verbose (bool, Optional): Whether to print verbose output. Defaults to False.
+        """
+
+        question_answer_definition: SeedPromptDataset = SeedPromptDataset.from_yaml_file(
+            question_answer_definition_path
+        )
+
+        self._user_start_turn = question_answer_definition.prompts[0]
+        self._assistant_start_turn = question_answer_definition.prompts[1]
+
+        super().__init__(
+            objective_target=objective_target,
+            scorers=scorers,
+            prompt_converters=prompt_converters,
+            verbose=verbose,
+            batch_size=batch_size,
+        )
+
+        self._set_default_conversation_start()
+
+    async def send_prompts_async(  # type: ignore[override]
+        self,
+        *,
+        dataset: QuestionAnsweringDataset,
+        prompt_type: PromptDataType = "text",
+        memory_labels: Optional[dict[str, str]] = None,
+        metadata: Optional[dict[str, Union[str, int]]] = None,
+    ) -> list[PromptRequestResponse]:
+        """Sends prompts to the chat model and evaluates responses."""
+        prompt_list = [construct_evaluation_prompt(entry) for entry in dataset.questions]
+        requests: list[NormalizerRequest] = []
+        for prompt in prompt_list:
+
+            requests.append(
+                self._create_normalizer_request(
+                    prompt_text=prompt,
+                    prompt_type=prompt_type,
+                    converters=self._prompt_converters,
+                    metadata=metadata,
+                    conversation_id=str(uuid.uuid4()),
+                )
+            )
+
+        return await self.send_normalizer_requests_async(
+            prompt_request_list=requests, memory_labels=memory_labels, dataset=dataset
+        )
```

I am not a fan of this, I understand the challenge here, but I think you can resolve it much better by using overloaded functions. That way, you do not need `# type: ignore[override]`. Also, a lot of code here is duplicated from the base class with a slight variation in the implementation which is not ideal. I would love to avoid that.

Here is an example how you could handle this without using `type ignore`
```python
   # Exactly matches the base signature
    @overload
    async def send_prompts_async(
        self,
        *,
        prompt_list: list[str],
        prompt_type: PromptDataType = ...,
        memory_labels: Optional[dict[str, str]] = ...,
        metadata: Optional[dict[str, Union[str, int]]] = ...,
    ) -> list[PromptRequestResponse]:
        ...
    
    # this is for your new dataset 
    @overload
    async def send_prompts_async(
        self,
        *,
        dataset: QuestionAnsweringDataset,
        prompt_type: PromptDataType = ...,
        memory_labels: Optional[dict[str, str]] = ...,
        metadata: Optional[dict[str, Union[str, int]]] = ...,
    ) -> list[PromptRequestResponse]:
        ...
       
    # and then you need to add the implementation this way
    async def send_prompts_async(
        self,
        *,
        prompt_list: Optional[list[str]] = None,
        dataset:     Optional[QuestionAnsweringDataset] = None,
        prompt_type: PromptDataType                = "text",
        memory_labels: Optional[dict[str, str]]     = None,
        metadata:      Optional[dict[str, Union[str, int]]] = None,
    ) -> list[PromptRequestResponse]:
        if dataset is not None:
            if prompt_list is not None:
                raise ValueError("Cannot pass both prompt_list and dataset")
            # convert your dataset into a list of strings
            prompt_list = [construct_evaluation_prompt(entry) for entry in dataset.questions]

        if len(prompt_list) == 0:
            raise ValueError( ... )

        # delegate to the base-class implementation
        return await super().send_prompts_async(
            prompt_list=prompt_list,
            prompt_type=prompt_type,
            memory_labels=memory_labels,
            metadata=metadata,
        )
```

The only thing you add to the `send_normalizer_requests_async` is if the scorer is an instance of `QuestionAnswerScorer`. Just because of that, the entire code is duplicated. 

Here is my recommendation:

1. Add the following method to the `PromptSendingOrchestrator`:
```python
def _score_responses(self, *, response_pieces: MutableSequence[PromptRequestPiece]) -> None:
      for scorer in self._scorers:
            await scorer.score_responses_inferring_tasks_batch_async(
                    request_responses=response_pieces, batch_size=self._batch_size
             )
```
2. Change the `send_normalizer_requests_async` in `PromptSendingOrchestrator` to use the new `_score_responses()`:
```Python
async def send_normalizer_requests_async(
        self,
        *,
        prompt_request_list: list[NormalizerRequest],
        memory_labels: Optional[dict[str, str]] = None,
    ) -> list[PromptRequestResponse]:
        """
        Sends the normalized prompts to the prompt target.
        """

        self.validate_normalizer_requests(prompt_request_list=prompt_request_list)

        for prompt in prompt_request_list:
            prompt.conversation_id = await self._prepare_conversation_async(normalizer_request=prompt)

        # Normalizer is responsible for storing the requests in memory
        # The labels parameter may allow me to stash class information for each kind of prompt.
        responses: list[PromptRequestResponse] = await self._prompt_normalizer.send_prompt_batch_to_target_async(
            requests=prompt_request_list,
            target=self._objective_target,
            labels=combine_dict(existing_dict=self._global_memory_labels, new_dict=memory_labels),
            orchestrator_identifier=self.get_identifier(),
            batch_size=self._batch_size,
        )

        if self._scorers and responses:
            response_pieces = PromptRequestResponse.flatten_to_prompt_request_pieces(responses)

            self._score_responses(response_pieces=response_pieces)

        return responses
```

3. Only override `_score_responses()` in `QuestionAnsweringBenchmarkOrchestrator`:
```Python
def _score_responses(self, *, response_pieces: MutableSequence[PromptRequestPiece]) -> None:
      for scorer in self._scorers:
                if isinstance(scorer, QuestionAnswerScorer):
                    await scorer.score_prompts_with_tasks_batch_async(
                        request_responses=response_pieces, batch_size=self._batch_size, tasks=dataset.questions
                    )
                else:
                    await scorer.score_responses_inferring_tasks_batch_async(
                        request_responses=response_pieces, batch_size=self._batch_size
                    )
```

This way, you avoid huge amount of code duplication

---

I understand but that could always be built into validations. If `prompt_list` is provided and not `dataset`, for example, you could throw a validation error. 
The way I suggested above cuts down the code duplication we are doing here. Also, the modification to `PromptSendingOrchestrator` is completely transparent from all consumers so it is not a behavior change.

---

The problem here is that we require additional context from the tasks being QuestionAnsweringEntry objects. We would need to sometimes pass tasks into _score_responses from send_normalizer_requests_async (like for this orchestrator) and other times tasks can be inferred. I think this means I would still need to write my own send_normalizer_requests_async. 

One way that I think would work is having tasks be a member variable of the orchestrator that is set when send_prompts_async() is called and then in the overloaded _score_responses I access that variable if needed.

---

ahh, I see what you mean, I missed that part ... I can see the issue now. If the tasks are only assessed during scoring, and they are different between batches, I would not recommend adding it to the scorer as a member, because that would add the state for the run directly in the scorer class. Think of how a multi-threaded execution would run into a problem!

I think in this case, we could use the type ignore, I think that's the least intrusive approach until we rearchitect the orchestrators.